lightning/restore/restore.go

```diff
@@ -1359,6 +1379,138 @@ func splitIntoDeliveryStreams(totalKVs []kvenc.KvPair, splitSize int) [][]kvenc.
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+type deliveredKVs struct {
+	kvs    []kvenc.KvPair // if kvs is empty, this indicated we've got the last message.
+	offset int64
+	rowID  int64
+}
+
+type deliverResult struct {
+	totalDur time.Duration
+	err      error
+}
+
+func (cr *chunkRestore) deliverLoop(
+	ctx context.Context,
+	kvsCh <-chan deliveredKVs,
+	t *TableRestore,
+	engineID int32,
+	dataEngine, indexEngine *kv.OpenedEngine,
+	rc *RestoreController,
+) (deliverTotalDur time.Duration, err error) {
+	var channelClosed bool
+	var dataKVs, indexKVs []kvenc.KvPair
+
+	for !channelClosed {
+		var dataChecksum, indexChecksum verify.KVChecksum
+		var offset, rowID int64
+
+		// Fetch enough KV pairs from the source.
+	populate:
+		for dataChecksum.SumSize()+indexChecksum.SumSize() < minDeliverBytes {
+			select {
+			case d := <-kvsCh:
+				if len(d.kvs) == 0 {
+					channelClosed = true
+					break populate
+				}
+
+				for _, kv := range d.kvs {
+					if kv.Key[tablecodec.TableSplitKeyLen+1] == 'r' {
+						dataKVs = append(dataKVs, kv)
+						dataChecksum.UpdateOne(kv)
+					} else {
+						indexKVs = append(indexKVs, kv)
+						indexChecksum.UpdateOne(kv)
+					}
+				}
+				offset = d.offset
+				rowID = d.rowID
+			case <-ctx.Done():
+				err = ctx.Err()
+				return
+			}
+		}
+
+		// Write KVs into the engine
+		start := time.Now()
+
+		if err = writeToEngine(ctx, dataEngine, dataKVs); err != nil {
+			return
+		}
+		if err = writeToEngine(ctx, indexEngine, indexKVs); err != nil {
+			return
+		}
+
+		deliverDur := time.Since(start)
+		deliverTotalDur += deliverDur
+		metric.BlockDeliverSecondsHistogram.Observe(deliverDur.Seconds())
+		metric.BlockDeliverBytesHistogram.WithLabelValues(metric.BlockDeliverKindData).Observe(float64(dataChecksum.SumSize()))
+		metric.BlockDeliverBytesHistogram.WithLabelValues(metric.BlockDeliverKindIndex).Observe(float64(indexChecksum.SumSize()))
+		metric.BlockDeliverKVPairsHistogram.WithLabelValues(metric.BlockDeliverKindData).Observe(float64(dataChecksum.SumKVS()))
+		metric.BlockDeliverKVPairsHistogram.WithLabelValues(metric.BlockDeliverKindIndex).Observe(float64(indexChecksum.SumKVS()))
+
+		dataKVs = dataKVs[:0]
+		indexKVs = indexKVs[:0]
+
+		// Update the table, and save a checkpoint.
+		// (the write to the importer is effective immediately, thus update these here)
+		// No need to apply a lock since this is the only thread updating these variables.
+		cr.chunk.Checksum.Add(&dataChecksum)
+		cr.chunk.Checksum.Add(&indexChecksum)
+		cr.chunk.Chunk.Offset = offset
+		cr.chunk.Chunk.PrevRowIDMax = rowID
+		if dataChecksum.SumKVS() != 0 || indexChecksum.SumKVS() != 0 {
+			// No need to save checkpoint if nothing was delivered.
+			cr.saveCheckpoint(t, engineID, rc)
+		}
+	}
+
+	return
+}
+
+func (cr *chunkRestore) saveCheckpoint(t *TableRestore, engineID int32, rc *RestoreController) {
+	rc.saveCpCh <- saveCp{
+		tableName: t.tableName,
+		merger: &RebaseCheckpointMerger{
```

Comment:
is it enough to only save it in L529?

Reply:
Unfortunately no, the AllocBase needs to be larger than all content of `_tidb_rowid` (or the integer primary key), which value cannot be determined until we've read all data.

Added a comment for this.